x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -882,6 +917,37 @@ public static Authentication newApiKeyAuthentication(AuthenticationResult<User>
         return authentication;
     }
 
+    public static Authentication newRemoteAccessAuthentication(
+        AuthenticationResult<User> remoteAccessApiKeyAuthcResult,
+        Authentication authenticationFromRemoteCluster,
+        String nodeName
+    ) {
+        assert remoteAccessApiKeyAuthcResult.isAuthenticated() : "Remote access authentication result must be successful";
+
+        final User remoteAccessApiKeyUser = remoteAccessApiKeyAuthcResult.getValue();
+        assert remoteAccessApiKeyUser.roles().length == 0
+            : "The user associated with a remote access API key authentication must have no role";
+
+        final User userFromRemoteCluster = authenticationFromRemoteCluster.getEffectiveSubject().getUser();
+        assert userFromRemoteCluster.enabled() : "The user received from a remote cluster must be enabled";
+
+        final User userWithoutRoles = new User(
+            userFromRemoteCluster.principal(),
```

Comment:
I decided against combining the API key ID with the remote user principal: simple concatenation would require somewhat weird edge-case handling around max length restrictions (the API key ID together with an arbitrary username could exceed the limit); hashing would have been expensive and made the principal opaque. 

In terms of what appears in the logs and API responses, the remote user principal seems more useful than the API key ID which is why I've chosen it here. Furthermore, the API key ID is included in the realm name.